actor-share-conversion/src/gf2_128/receiver.rs

```diff
@@ -0,0 +1,233 @@
+            .map_err(|err| ShareConversionError::Other(err.to_string()))?
+    }
+}
+
+#[async_trait]
+impl<T> VerifyTape for ReceiverControl<T>
+where
+    T: Handler<VerifyTapeMessage, Return = Result<(), ShareConversionError>>,
+{
+    async fn verify_tape(self) -> Result<(), ShareConversionError> {
+        self.0
+            .send(VerifyTapeMessage)
+            .await
+            .map_err(|err| ShareConversionError::Other(err.to_string()))?
+    }
+}
+
+#[async_trait]
+impl<T, OT, U, V, W> Handler<M2AMessage<Vec<u128>>> for Receiver<T, OT, U, V, W>
+where
+    T: AsyncFactory<OT> + Send + 'static,
+    OT: ObliviousReceive<bool, Block> + Send + 'static,
+    U: Gf2_128ShareConvert + Send + 'static,
+    V: MuxChannelControl<Gf2ConversionMessage> + Send + 'static,
+    W: Recorder<U> + Send + 'static,
+    IOReceiver<T, OT, U, W>: MultiplicativeToAdditive<FieldElement = u128>,
+{
+    type Return = Result<Vec<u128>, ShareConversionError>;
+
+    async fn handle(
+        &mut self,
+        message: M2AMessage<Vec<u128>>,
+        _ctx: &mut Context<Self>,
+    ) -> Self::Return {
+        match self.state {
+            State::Setup(ref mut state) => state.m_to_a(message.0).await,
+            _ => Err(ShareConversionError::Other(String::from(
+                "Actor is not setup",
+            ))),
+        }
+    }
+}
+
+#[async_trait]
+impl<T, OT, U, V, W> Handler<A2MMessage<Vec<u128>>> for Receiver<T, OT, U, V, W>
+where
+    T: AsyncFactory<OT> + Send + 'static,
+    OT: ObliviousReceive<bool, Block> + Send + 'static,
+    U: Gf2_128ShareConvert + Send + 'static,
+    V: MuxChannelControl<Gf2ConversionMessage> + Send + 'static,
+    W: Recorder<U> + Send + 'static,
+    IOReceiver<T, OT, U, W>: AdditiveToMultiplicative<FieldElement = u128>,
+{
+    type Return = Result<Vec<u128>, ShareConversionError>;
+
+    async fn handle(
+        &mut self,
+        message: A2MMessage<Vec<u128>>,
+        _ctx: &mut Context<Self>,
+    ) -> Self::Return {
+        match self.state {
+            State::Setup(ref mut state) => state.a_to_m(message.0).await,
+            _ => Err(ShareConversionError::Other(String::from(
+                "Actor is not setup",
+            ))),
+        }
+    }
+}
+
+#[async_trait]
+impl<T, OT, U, V> Handler<VerifyTapeMessage> for Receiver<T, OT, U, V, Tape>
+where
+    T: AsyncFactory<OT> + Send + 'static,
+    OT: ObliviousReceive<bool, Block> + Send + 'static,
+    U: Gf2_128ShareConvert + Send + 'static,
+    V: MuxChannelControl<Gf2ConversionMessage> + Send + 'static,
+    IOReceiver<T, OT, U, Tape>: VerifyTape,
+{
+    type Return = Result<(), ShareConversionError>;
+
+    async fn handle(
+        &mut self,
+        _message: VerifyTapeMessage,
+        ctx: &mut Context<Self>,
+    ) -> Self::Return {
+        let state = std::mem::replace(&mut self.state, State::Complete);
+
+        let State::Setup(state) = state else {
+            return Err(ShareConversionError::Other(String::from(
```

Comment:
the state should be restored just like my comment in sender.rs

Reply:
same as clarification above, the reverting should happen before the line
`return Err(ShareConversionError::Other(String::from(
                "Actor is not setup"
            )));`